docs/reference/speed_python.rst

```diff
@@ -80,6 +80,19 @@ point to sections of the code where performance is not paramount. For example,
 capture ADC readings as integers values to an array in one quick go, and only then
 convert them to floating-point numbers for signal processing.
 
+Integers
+~~~~~~~~
+
+On most ports, integers between -2**30 and 2**30-1 are treated as "small
+integers", and the value is stored directly in the object handle.  For integers
+outside this range, the value is stored in space allocated from the heap,
+resulting in heap allocations and consequent garbage collection when such
+values are copied or passed to functions.
```

Comment:
Thanks @pdw-mb this is a very useful addition!

Just a note though -- passing to functions doesn't cause an allocation, and "copy" is a bit of a loose concept here. Really the only operation that causes allocation is arithmetic operations on the integers.

To clarify:

```py
a = ... # some large int value
b = a # no allocation
c = a + 1 # allocates
some_func(a, b, c)  # no allocation
d = a - a  # allocates, even though the result is a small int (zero)
```

For more about that last point, see https://github.com/micropython/micropython/pull/9531

Reply:
Thank you for the feedback.   I think I now understand this better and will attempt to re-word.

The specific thing that tripped me up was:

```py
some_func(a[n])
```
where `a` is an array of type `I` (see https://github.com/micropython/micropython/discussions/10475)

So it seems to me that allocation happens when a _new_ integer value is obtained, either by calculation, or by conversion from another format (e.g. from an array element).  Presumably an allocation wouldn't occur for the above if `a` were a list as the elements are already Python objects?

Reply:
> Presumably an allocation wouldn't occur for the above if `a` were a list as the elements are already Python objects?

Correct.

Reply:
Now reworded to:

> For integers outside this range, the value is stored in space allocated from the heap,
resulting in heap allocations, and consequent garbage collection, when such values are instantiated.  This occurs where values are calculated from other values, or converted from other formats such as an array.

Reply:
Shouldn't that be 'this can for instance occur when ....' ? Because what you mean is that `a*b` of `int(<somefloat>)` can (but not necessarily will) result in an alloction, right?
Also: 'value converted from array' is not really clear: what kind of array? Is that something common enough to add here?

Reply:
> Because what you mean is that a*b of int(<somefloat>) can (but not necessarily will) result in an alloction, right?

My understanding from @jimmo's comment was that if either `a` or `b` is a large int value it _will_ allocate.  It should probably say where _such_ values are calculated.  Maybe I should include @jimmo's examples as @robert-hh suggests.

> value converted from array

Specifically:

```
arr = array.array('I', [2**10, 2**31])
a[0] # doesn't allocate
a[1] # allocates
```

This is the specific thing that bit me (and it wasn't my code, so the original author clearly missed it too).  I'm happy to make it more explicit, but I do think it's worth mentioning.

Reply:
> Specifically:

Right, but that has nothing do to wit the value coming from an array, it's just that the number happens to be larger than what fits, right?